src/Composer/Package/Cleaner/Cleaner.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace LastCall\DownloadsPlugin\Composer\Package\Cleaner;
+
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\Gitignore;
+
+class Cleaner implements CleanerInterface
+{
+    public function __construct(private array $ignore)
+    {
+    }
+
+    public function clean(string $dir): void
+    {
+        if (empty($this->ignore)) {
+            return;
+        }
+
+        $finder = new Finder();
+        $finder->in($dir)->path(Gitignore::toRegex(implode(\PHP_EOL, $this->ignore)));
+
+        $dirs = [];
+        foreach ($finder as $item) {
+            if ($item->isDir()) {
+                $dirs[] = $item->getPathname();
+            } else {
+                unlink($item->getPathname());
+            }
+        }
+        foreach ($dirs as $dir) {
+            @rmdir($dir);
+        }
```

Comment:
I wonder if there is a hook we could use to do this, rather than an inheritance.

Composer has such event hooks, doesn't it?

Reply:
Not meant as a blocker, more a discussion point

Reply:
The current flow is:
1. package installed
2. POST_PACKAGE_INSTALL event
3. create extra file VIRTUAL packages
4. Run INSTALL operation for these virtual packages
  1. Download all files at once
  2. Prepare
  3. Install -> promise -> then **clean ignored files**
  4. IMPORTANT -> don't trigger script events
  5. Clean up (do nothing)

Currently I disable .4.4 because of the infinite loop. 

I can try it if needed.

I can also try to move the code from .4.3 to .4.5

Reply:
Ah I see, script events could cause infinite loop. Probably worth calling out in README.md so that if this messes with other composer plugins, folks know why.

Reply:
Currently I don't know what to add to README.md, because I can't imagine exact situation when this plugin conflict with other plugins. Can you help me with that?